Migrate airflow-importing-custom-hooks-operators guide (#1077)
astronomer/docs: guides/airflow-importing-custom-hooks-operators.md

---
title: "Import custom hooks and operators"
sidebar_label: "Import custom hooks and operators"
description: "How to correctly import custom hooks and operators."
id: airflow-importing-custom-hooks-operators

> tags: ["Hooks", "Operators", "Plugins", "Basics"]

Reviewer: ```suggestion
```

from airflow.utils.decorators import apply_defaults
from hooks.my_hook import MyHook

class MyOperator(BaseOperator):

    @apply_defaults
    def __init__(self,
                 my_field,
                 *args,
                 **kwargs):
        super(MyOperator, self).__init__(*args, **kwargs)
        self.my_field = my_field

    def execute(self, context):
        hook = MyHook('my_conn')
        hook.my_method()
```

If your custom operator is modifying functionality of an existing operator, your class can inherit from the operator you are building on instead of the `BaseOperator`. For more detailed instructions on defining custom operators, see the [Apache Airflow How-to Guide](https://airflow.apache.org/docs/apache-airflow/stable/howto/custom-operator.html).

## Import custom operators

After you've defined your custom operator, you need to make it available to your DAGs. Some legacy Airflow documentation or forums may reference registering your custom operator as an Airflow plugin, but this is not necessary. To import  a custom operator into your DAGs, the file containing your custom operator needs to be in a directory that is present in your `PYTHONPATH`.

> Airflow by default will add the `dags/` and `plugins/` directories in a project to the `PYTHONPATH`, so those are the most natural choices for storing custom operator files (check out the Apache Airflow [Module Management docs](https://airflow.apache.org/docs/apache-airflow/stable/modules_management.html) for more info). Your project structure may vary depending on your team and your use case. Astronomer uses the following structure, where custom operator files live in the `plugins/` directory with sub-folders for readability.

Reviewer: ```suggestion
By default, Airflow adds the `dags/` and `plugins/` directories in a project to the `PYTHONPATH`, so those are the most natural choices for storing custom operator files (check out the Apache Airflow [Module Management docs](https://airflow.apache.org/docs/apache-airflow/stable/modules_management.html) for more info). Your project structure may vary depending on your team and your use case. Astronomer uses the following structure, where custom operator files live in the `plugins/` directory with sub-folders for readability.
```

```bash
.
├── dags/                    
│   ├── example-dag.py
├── Dockerfile                  
├── include/                 
│   └── sql/
│       └── transforms.sql
├── packages.txt     
├── plugins/             
│   └── operators/
│       └── my_operator.py
│   └── sensors/
│       └── my_sensor.py
└── requirements.txt    
```

For more details on why Astronomer recommends this project structure, see the [Managing Airflow Code guide](https://www.astronomer.io/guides/managing-airflow-code).

> > Note: If you use an IDE and don't want to see import errors, add the `plugins` directory as a source root.

Reviewer: ```suggestion
:::tip

If you use an IDE and don't want to see import errors, add the `plugins` directory as a source root.

:::
```
